### DISCUSSION TITLE
Re-training exisitng NER trf model

### DISCUSSION THREAD

**Discussion by GladiatorX** [explosion/spaCy#9078] [Help: Coding & Implementations] [answered]
I am trying to update an existing spacy model**en_core_web_trf** for **DATE entity** with some additional date format's.

The data is properly formatted and converted to **.spacy** file as mentioned in spacy's documentation

In **config.cfg** file as per the [link](https://spacy.io/usage/processing-pipelines#sourced-components)
```
[components.ner]
source = "en_core_web_trf"
```
The trained model results in catastrophic forgetting issue:
- Previously detected some formats gets misclassified
- The updated en_core_web_trf model is unable to determine any other exiting entity like PERSON, ORDINAL, LOC etc..(even though ```nlp.components[5][1].labels``` shows all existing entity list )

How to deal with this issue without compromising on affecting default accuracy ? Does it need some additional parameters to go into **config.cfg** file?

**Comment by polm** [comment, marked as answer]
The best way to avoid catastrophic forgetting is to train with a complete dataset. As an alternate option, you can use the "rehearsal" APIs, but they're experimental and probably need some work to use them at all. 

Since you're only interested in changing dates, I'd recommend training an NER model that just marks dates and putting it after the main model. If you specifically also want to ignore dates from the main model, you can make a custom component to erase DATE annotations first. 

Also see the [FAQ section on catastrophic forgetting](https://github.com/explosion/spaCy/discussions/8226#:~:text=My%20retrained%20model,7666%20(comment)).

**Reply by DarkSoliditi** [reply]
Where can I find documentation on erasing existing annotations and and replacing them with a custom component. DATE is something I may have to retrain as I will need the model to only recognize complete dates and ignore partials.

**Reply by polm** [reply]
There are a couple of ways to erase annotations.

1. You can set `ent_type_` to `""` and `ent_iob_` to `O` on the affected tokens
2. You can filter `doc.ents` to the list of entities you want to keep and set `doc.ents` (this may be easier)

For some explanation of how multiple NER-labelling components work you can see the [double NER tutorial project](https://github.com/explosion/projects/tree/v3/tutorials/ner_double).

**Reply by GladiatorX** [reply]
I like your idea of having 2nd model so I trained it on additional kinds of date formats that I would like it to detect, I set **_DT_** as the new label.. 
The trained model Infact detecting new matches but flip side is that it also detects some partial dates which the base_trf model has correctly identified.. ( this seems obvious considering how nature of data the dates have)
The behavior is spurious. 

It would have been great we can freeze the `base_trf `model and just train last few layers  with newer instances/data  for new entity label.

**Reply by polm** [reply]
In theory you can use the recently added `annotating_components` to freeze the input layer and just train an NER on top of it, though I think it may be having issues with Transformers at the moment.

**Comment by DarkSoliditi** [comment]
You always come through Paul! Thanks again!

**Reply by GladiatorX** [reply]
@DarkSoliditi You can try passing the dates identified by spacy NER to **dateparser** package, set parameter `strict=True` this should give desired output
